File: src/main/java/com/ddd/trainings/api/app/TrainingIdeaService.java
package com.ddd.trainings.api.app;

import com.ddd.trainings.domain.IdeaId;
import com.ddd.trainings.domain.ProposalId;
import com.ddd.trainings.domain.TrainingIdea;
import com.ddd.trainings.domain.TrainingIdeaFactory;
import com.ddd.trainings.domain.TrainingProposal;
import com.ddd.trainings.infra.TrainingIdeaRepo;
import com.ddd.trainings.infra.TrainingProposalRepo;
import java.util.Optional;
import org.jmolecules.ddd.annotation.Service;

@Service
public class TrainingIdeaService {
  
  private final TrainingIdeaRepo ideaRepo;
  private final TrainingProposalRepo proposalRepo;

  public TrainingIdeaService(TrainingIdeaRepo ideaRepo, TrainingProposalRepo proposalRepo) {
    this.ideaRepo = ideaRepo;
    this.proposalRepo = proposalRepo;
  }

  public IdeaId createTrainingIdea(String trainerId, String name) {
    TrainingIdea idea = TrainingIdeaFactory.create(trainerId, name);
    TrainingIdea saved = ideaRepo.save(idea);
    return saved.getId();
  }

  public ProposalId proposeIdea(String ideaId) {
    Optional<TrainingIdea> idea = ideaRepo.findById(IdeaId.from(ideaId));
    if(idea.isPresent()) {
      TrainingProposal proposal = idea.get().propose();
      proposalRepo.save(proposal);
      return proposal.getId();
    } else {
      throw new RuntimeException("Idea not found");

    }
  }
}

File: src/Ddd/Trainings/Idea/Api/App/TrainingIdeaService.cs
using Ddd.Trainings.Idea.Domain;
using Ddd.Trainings.Idea.Infra;
using Ddd.Trainings.Proposal.Domain;
using Ddd.Trainings.Proposal.Infra;

namespace Ddd.Trainings.Idea.Api.App
{
    public class TrainingIdeaService(ITrainingIdeaRepo ideaRepo, TrainingProposalRepo proposalRepo, ReviewerPolicyFactory reviewerPolicyFactory)
  {
        private readonly ITrainingIdeaRepo _ideaRepo = ideaRepo;
        private readonly TrainingProposalRepo _proposalRepo = proposalRepo;
        private readonly ReviewerPolicyFactory _reviewerPolicyFactory = reviewerPolicyFactory;

    public IdeaId CreateTrainingIdea(string trainerId, string name)
        {
            TrainingIdea idea = TrainingIdeaFactory.Create(trainerId, name);
            TrainingIdea saved = _ideaRepo.Save(idea);
            return saved.GetId();
        }

        public ProposalId ProposeIdea(string ideaId)
        {
            TrainingIdea idea = _ideaRepo.FindById(IdeaId.From(ideaId)) ?? throw new Exception("Idea not found");
            TrainingProposal proposal = idea.Propose(_reviewerPolicyFactory.ByKeywords(idea.Name));
            _proposalRepo.Save(proposal);
            return proposal.Id;
        }
    }
} 